src/core/experimental/rc/slice.d

```diff
@@ -0,0 +1,217 @@
+
+    /**
+    Creates a new `__rcslice` instance, allocating memory for the slice.
+
+    Params:
+         size = size of the slice, in number of elements of type `T`
+    */
+    this(size_t size) inout
+    {
+        import core.memory : pureMalloc;
+
+        auto memory = pureMalloc(CounterType.sizeof + T.sizeof * size);
+
+        this.slice = (() @trusted inout => (cast(inout T*) (memory + CounterType.sizeof))[0 .. size])();
+
+        this.count = (() @trusted inout => cast(typeof(count)) memory)();
+        cast() *this.count = 0;
+
+        static if (defaultInitialize)
+        {
+            import core.internal.lifetime : emplaceRef;
+            foreach (i; 0 .. size)
+            {
+                () @trusted { emplaceRef((cast(T*) slice.ptr)[i]); }();
+            }
+        }
+    }
+
+    ///
+    inout(T[]) opSlice(size_t start, size_t end) inout return scope
+    {
+        return slice[start .. end];
+    }
+
+    ///
+    ref inout(T[]) opSlice() inout return scope
+    {
+        return slice;
+    }
+
+    ///
+    ref inout(T) opIndex(size_t i) inout return scope
+    {
+        return slice[i];
+    }
+
+    ///
+    inout(T*) ptr() inout return scope
+    {
+        return slice.ptr;
+    }
+
+    ///
+    size_t length() inout
+    {
+        return slice.length;
+    }
+
+    ///
+    bool opEquals(typeof(null)) inout
+    {
+        return slice is null;
+    }
+
+    ~this()
+    {
+        delRef();
+    }
+
+    ///
+    void opAssign(typeof(this) rhs)
+    {
+        if (rhs.count == count)
+        {
+            return;
+        }
+
+        //TODO: Move std.algorithm.mutation.swap to druntime
+        auto tmp1 = rhs.slice;
+        rhs.slice = slice;
+        slice = tmp1;
+
+        auto tmp2 = rhs.count;
+        rhs.count = count;
+        count = tmp2;
+    }
+
+    ///
+    void opAssign(typeof(null))
+    {
+        delRef();
+    }
+
+    ///
+    this(scope ref inout typeof(this) rhs) inout
+    {
+        slice = rhs.slice;
+        count = rhs.count;
+
+        addRef();
+    }
+
+    private void addRef() inout
+    {
+        import core.atomic : atomicOp;
+
+        if (slice is null)
+        {
+            return;
+        }
+
+        //TODO: Use `atomicFetchAdd` with `MemoryOrder.seq` once LDC has it
+        () @trusted { atomicOp!"+="(*(cast(shared(CounterType)*) count), 1); } ();
+    }
+
+    private void delRef()
+    {
+        import core.atomic : atomicLoad, atomicOp, MemoryOrder;
+
+        if (slice is null)
+        {
+            return;
+        }
+
+        // The counter is left at -1 when this was the last reference
+        // (i.e. the counter is 0-based, because we use calloc)
+        //TODO: Use `atomicFetchSub` with `MemoryOrder.raw` once LDC has it
+        if (atomicLoad!(MemoryOrder.raw)(*count) == 0 || atomicOp!("-=")(*count, 1) == -1)
+        {
+            deallocate();
+        }
+    }
+
+    @trusted
+    private void deallocate()
```

Comment:
Strictly speaking, this can't be `@trusted`. It's essentially an alias of `free`, which can't possibly be `@trusted`.

One might argue that it's ok here, because it's private and the rest of the code takes care to call it correctly. But that's prone to the kind of mistakes that `@safe` is supposed to prevent.

And there already is such a mistake in the code: `opAssign(null)` is public and calls `delRef` which calls `deallocate`. We effectively have a public `@trusted` alias of `free`, and we can use it to create dangling references in `@safe` code:

```d
auto rcs = __rcslice!int(1);
int[] s = rcs[];
rcs = null; /* leaves s dangling */
```

Reply:
Thanks for pointing this out!